### PR TITLE
survey checks

### PR DESCRIPTION
**Pull request by kzidane**
How does this look as a start?

### CODE REVIEW

**Comment by brianyu28** [comment]
@kzidane Do you think there's any possible way we can check that `/sheet` is (likely) working?

I'm thinking maybe...

1. `GET /sheet` save the response.
2. Look through `form.html` for any `input` fields and get their `name`s
3. POST some dummy data to `/form` with those `name`s
4. `GET /sheet` again and check at minimum that the returned HTML is distinct (but in the best case, check that the POSTed dummy data is present in the table)

Think that's doable? Anything I'm missing in that thinking?

File: cs50/2018/fall/survey/check50/__init__.py [hunk 1]
```diff
@@ -0,0 +1,71 @@
...
+    @check()
+    def exists(self):
+        """application.py exists"""
+        self.require("application.py")
+
+
+    @check("exists")
+    def startup(self):
+        """application starts up"""
+        self.flask(Survey.APP).get("/").status(200)
```
Comment (dmalan) [inline cs50/2018/fall/survey/check50/__init__.py:34]: `/` actually returns a 302 by design.

File: cs50/2018/fall/survey/check50/__init__.py [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+
+
+    @check("startup")
+    def has_form(self):
+        """has form"""
+
+        if len(self.get_content("/", "form")) != 1:
```
Comment (dmalan) [inline cs50/2018/fall/survey/check50/__init__.py]: `/form` has the content.

Comment (dmalan) [reply cs50/2018/fall/survey/check50/__init__.py]: Can probably check for `== 0` instead to allow for multiple for whatever reason.

File: cs50/2018/fall/survey/check50/__init__.py [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+            raise Error("expected form element")
+
+
+    @check("has_form")
+    def has_text_fields(self):
+        """has one or more text fields"""
+
+        inputs = self.get_content("/", "form")[0].find_all("input")
```
Comment (dmalan) [inline cs50/2018/fall/survey/check50/__init__.py]: Can you search directly via bs for `input[type=text]` or such?

File: cs50/2018/fall/survey/check50/__init__.py [hunk 4]
```diff
@@ -0,0 +1,71 @@
...
+        if len(tuple(filter(lambda e: e.attrs["type"] == "text", inputs))) < 1:
+            raise Error("expected at least one text input")
```
Comment (dmalan) [inline cs50/2018/fall/survey/check50/__init__.py]: Should check for `textarea` too as an alternative. And also `type="email"`, `password`, etc.

File: cs50/2018/fall/survey/check50/__init__.py [hunk 5]
```diff
@@ -0,0 +1,71 @@
...
+
+
+    @check("has_form")
+    def has_checkbox_or_radio_buttons(self):
+        """has one or more checkboxes or two or more radio buttons"""
+
+        inputs = self.get_content("/", "form")[0].find_all("input")
+        if len(tuple(filter(lambda e: e.attrs["type"] == "checkbox", inputs))) < 1 and \
+            len(tuple(filter(lambda e: e.attrs["type"] == "radio", inputs))) < 2:
+            raise Error("expected at least one checkbox or two radio buttons")
```
Comment (dmalan) [inline cs50/2018/fall/survey/check50/__init__.py:72]: nice

File: cs50/2018/fall/survey/check50/__init__.py [hunk 6]
```diff
@@ -0,0 +1,71 @@
...
+
+
+    @check("has_form")
+    def has_select_and_options(self):
+        """has one or more select menus, each with two or more options"""
+
+        selects = self.get_content("/", "form")[0].find_all("select")
+        if len(selects) < 1 or \
+            any((len(select.find_all("option")) < 2 for select in selects)):
+            raise Error("expected one or more select menus, each with two or more options")
```
Comment (dmalan) [inline cs50/2018/fall/survey/check50/__init__.py:82]: nice

**Review by dmalan — CHANGES_REQUESTED** [review]

File: cs50/2018/fall/survey/check50/__init__.py (hunk 1 above)
Comment (brianyu28) [reply cs50/2018/fall/survey/check50/__init__.py:34]: `check50`'s `get` method by default uses `follow_redirects=True`, so this is really just testing that after following all redirects, the resulting page's status code is 200, so I think this should work.

**Comment by kzidane** [comment]
@brianyu28 per email, we could indeed try posting dummy data and grepping through content for arbitrary pieces of what we posted. One problem is if they expect the data to be in specific formats and do validations on backend we might not get what we expect. Not sure how reliable or accurate that would be tbh.

**Pull request merged**